src/main/java/fact/extraction/SinglePulseExtraction.java

```diff
@@ -57,47 +58,47 @@ public Data process(Data input) {
 
         npix = (Integer) input.get("NPIX");
         roi = (Integer) input.get("NROI");
-        double[] timeLines = (double[]) input.get(dataKey);
+        double[] timeSerieses = (double[]) input.get(dataKey);
 
-        double[] single_pe_count = new double[npix];
+        double[] numberOfPulses = new double[npix];
         int[][] pixelArrivalSlices = new int[npix][];
+        double[] baseLine = new double[npix];
+        double[][] timeSeriesAfterExtraction = new double[npix][];
 
-        double[] reducedTimeline = new double[timeLines.length];
+        SinglePulseExtractor.Config config = new SinglePulseExtractor.Config();
+        config.maxIterations = maxIterations;
+        SinglePulseExtractor spe = new SinglePulseExtractor(config);  
 
         for (int pix = 0; pix < npix; pix++) {
             int start = pix*roi+startSlice;
             int end = start + windowLength;
 
-            double[] pixelTimeLineInMv = Arrays.copyOfRange(
-                timeLines,
+            double[] pixelTimeSeriesInMv = Arrays.copyOfRange(
+                timeSerieses,
                 start, 
                 end
             );
             
-            double[] pixelTimeLine = SinglePulseExtractor. 
-                milliVoltToNormalizedSinglePulse(pixelTimeLineInMv);
-
-            int[] arrivalSlices = SinglePulseExtractor.
-                getArrivalSlicesOnTimeline(
-                    pixelTimeLine,
-                    maxIterations
-                );
-
-            single_pe_count[pix] = arrivalSlices.length;
-            pixelArrivalSlices[pix] = arrivalSlices;
-
-            for (int i = 0; i < windowLength; i++) {
-                reducedTimeline[start+i] = SinglePulseExtractor.factSinglePeAmplitudeInMv*pixelTimeLine[i];
-            }
+            double[] pixelTimeSeries = ElementWise.multiply(
+                pixelTimeSeriesInMv, 1.0/config.factSinglePeAmplitudeInMv);
+
+            SinglePulseExtractor.Result result = spe.extractFromTimeSeries(
+                pixelTimeSeries);
+
+            numberOfPulses[pix] = result.numberOfPulses();
+            pixelArrivalSlices[pix] = result.pulseArrivalSlices;
+            timeSeriesAfterExtraction[pix] = ElementWise.multiply(
+                result.timeSeriesAfterExtraction, 
+                config.factSinglePeAmplitudeInMv);
+            baseLine[pix] = result.timeSeriesBaseLine();
         }
 
-
-
         addStartSliceOffset(pixelArrivalSlices);
-        // printArrivalSlices(pixelArrivalSlices);
+
         input.put(outputKey, pixelArrivalSlices);
-        input.put(outputKey+"TL", reducedTimeline);
-        input.put(outputKey+"Count", single_pe_count);
+        input.put(outputKey+"TimeSeriesAfterExtraction", flatten(timeSeriesAfterExtraction));
```

Comment:
we could start to use the naming convention from the fact_tools refactoring branch e.g. outputkey+":TimeSeriesAfterExtraction".